### PR TITLE
Introducing OpenApi Generator Extension

### LINKED ISSUE #16401: OpenAPI generator for Quarkus REST Client
ATM there doesn't seem to be any way to generate REST client stubs from openapi for Quarkus, leading some of our users to use the following combination of pom.xml:

```xml
            <plugin>
                <groupId>org.openapitools</groupId>
                <artifactId>openapi-generator-maven-plugin</artifactId>
                <version>${openapi-generator-maven-plugin.version}</version>
                <executions>
                    <execution>
                        <id>generate-model</id>
                        <goals>
                            <goal>generate</goal>
                        </goals>
                        <configuration>
                            <library>resteasy</library>
                            <generateModels>true</generateModels>
                            <generateApis>false</generateApis>
                        </configuration>
                    </execution>
                    <execution>
                        <id>generate-api</id>
                        <goals>
                            <goal>generate</goal>
                        </goals>
                        <configuration>
                            <library>microprofile</library>
                            <generateModels>false</generateModels>
                            <generateApis>true</generateApis>
                            <generateSupportingFiles>true</generateSupportingFiles>
                            <templateDirectory>${project.basedir}/openapi-generator/templates</templateDirectory>
                        </configuration>
                    </execution>
                </executions>
            </plugin>

```

and template:

```
{{>licenseInfo}}
package {{package}};

{{#imports}}import {{import}};
{{/imports}}

import java.io.InputStream;
import java.io.OutputStream;
import java.util.List;
import java.util.Map;

import javax.enterprise.context.ApplicationScoped;
import javax.ws.rs.*;
import javax.ws.rs.core.Response;
import javax.ws.rs.core.MediaType;
{{^disableMultipart}}
{{! commented to disable cxf dependency,
otherwise override the multipart template file formParams.mustache to use RestEasy
import org.apache.cxf.jaxrs.ext.multipart.*;
}}
{{/disableMultipart}}

import org.eclipse.microprofile.rest.client.annotation.RegisterProvider;
import org.eclipse.microprofile.rest.client.inject.RegisterRestClient;

{{#appName}}
/**
 * {{{appName}}}
 *
 {{#appDescription}}
 * <p>{{{appDescription}}}
 *
 {{/appDescription}}
 */
{{/appName}}

@ApplicationScoped
@RegisterRestClient(configKey="{{classVarName}}")
@RegisterProvider(ApiExceptionMapper.class)
@Path("{{#useAnnotatedBasePath}}{{contextPath}}{{/useAnnotatedBasePath}}{{commonPath}}")
public interface {{classname}}  {
{{#operations}}
{{#operation}}

    {{#summary}}
    /**
     * {{summary}}
     *
     {{#notes}}
     * {{notes}}
     *
     {{/notes}}
     */
    {{/summary}}
    @{{httpMethod}}
{{#subresourceOperation}}
    @Path("{{{path}}}")
{{/subresourceOperation}}
{{#hasConsumes}}
    @Consumes({ {{#consumes}}"{{{mediaType}}}"{{^-last}}, {{/-last}}{{/consumes}} })
{{/hasConsumes}}
{{#hasProduces}}
    @Produces({ {{#produces}}"{{{mediaType}}}"{{^-last}}, {{/-last}}{{/produces}} })
{{/hasProduces}}
    public {{#returnType}}{{{returnType}}}{{/returnType}}{{^returnType}}void{{/returnType}} {{nickname}}({{#allParams}}{{>queryParams}}{{>pathParams}}{{>headerParams}}{{>bodyParams}}{{>formParams}}{{^-last}}, {{/-last}}{{/allParams}}) throws ApiException, ProcessingException;
{{/operation}}
}
{{/operations}}
```

Now, perhaps I'm wrong and there's a better way to generate the client stubs. @phillip-kruger @michalszynkiewicz @geoand do you know?

If there isn't, perhaps we should provide our own upstream, or write our own plugin for that, with Qute instead of this (@mkouba ) and with a custom version for RESTEasy Reactive too?

### PR DESCRIPTION
**Pull request by ricardozanini**
Draft to close #16401 

This is a **raw** introduction for the extension to-be.

### CODE REVIEW

**Comment by kenfinnigan** [comment]
Have you spoken with @cescoffier? I know he was looking at lots of options around this

**Comment by ricardozanini** [comment]
> Have you spoken with @cescoffier? I know he was looking at lots of options around this

Hi @kenfinnigan! I have not. I guess I'll talk to him first before proceeding.

File: extensions/openapi-generator/deployment/pom.xml [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+        </exclusion>
+        <exclusion>
+          <groupId>commons-cli</groupId>
+          <artifactId>commons-cli</artifactId>
+        </exclusion>
+      </exclusions>
+    </dependency>
+    <dependency>
+      <groupId>io.quarkus</groupId>
+      <artifactId>quarkus-qute-deployment</artifactId>
```
Comment (cescoffier) [inline extensions/openapi-generator/deployment/pom.xml:42]: I like the idea of using qute!

File: extensions/openapi-generator/deployment/pom.xml [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+  <artifactId>quarkus-openapi-generator-deployment</artifactId>
+  <name>Quarkus - OpenApi Generator - Deployment</name>
+
+  <properties>
+    <version.org.openapitools>5.1.1</version.org.openapitools>
+  </properties>
+
+  <dependencies>
+    <!-- TODO: move to BOM -->
```
Comment (cescoffier) [inline extensions/openapi-generator/deployment/pom.xml:20]: As said in the comment - these need to be moved to the bom.

File: extensions/openapi-generator/deployment/src/test/resources/templates/hello.qute [hunk 3]
```diff
@@ -0,0 +1 @@
+Hello! My name is {name}
```
Comment (cescoffier) [inline extensions/openapi-generator/deployment/src/test/resources/templates/hello.qute:1]: is this used?

Comment (ricardozanini) [reply extensions/openapi-generator/deployment/src/test/resources/templates/hello.qute:1]: Just for the initial "test first/validate the templating engine". I'll change it in the final version.

**Comment by ricardozanini** [comment]
This is how the client is being generated:

```java
package org.openapitools.client.api;

import org.openapitools.client.model.Order;

import org.eclipse.microprofile.rest.client.inject.RegisterRestClient;

import java.io.InputStream;
import java.io.OutputStream;
import java.util.List;
import java.util.Map;
import javax.ws.rs.*;
import javax.ws.rs.core.Response;
import javax.ws.rs.core.MediaType;
import javax.ws.rs.Path;
import org.jboss.resteasy.annotations.jaxrs.*;

import org.apache.cxf.jaxrs.ext.multipart.*;

/**
  * Swagger Petstore - OpenAPI 3.0
  * <p/>
  * This is a sample Pet Store Server based on the OpenAPI 3.0 specification.  You can find out more about Swagger at [http://swagger.io](http://swagger.io). In the third iteration of the pet store, we've switched to the design first approach! You can now help us improve the API whether it's by making changes to the definition itself or to the code. That way, with time, we can improve the API in general, and expose some of the new features in OAS3.  Some useful links: - [The Pet Store repository](https://github.com/swagger-api/swagger-petstore) - [The source API definition for the Pet Store](https://github.com/swagger-api/swagger-petstore/blob/master/src/main/resources/openapi.yaml)
  */
@Path("/store")
@RegisterRestClient
public interface StoreApi {

    @DELETE
    @Path("/order/{orderId}")
    public void deleteOrder(@PathParam("orderId") Long orderId);

    @GET
    @Path("/inventory")
    @Produces({"application/json"})
    public Map<String, Integer> getInventory();

    @GET
    @Path("/order/{orderId}")
    @Produces({"application/xml", "application/json"})
    public Order getOrderById(@PathParam("orderId") Long orderId);

    @POST
    @Path("/order")
    @Consumes({"application/json", "application/xml", "application/x-www-form-urlencoded"})
    @Produces({"application/json"})
    public Order placeOrder(Order order);

}
```

It's already possible to play around with it :)

File: extensions/openapi-generator/deployment/src/main/java/io/quarkus/openapi/generator/deployment/OpenApiGeneratorExtensionProcessor.java [hunk 4]
```diff
@@ -0,0 +1,15 @@
+package io.quarkus.openapi.generator.deployment;
+
+import io.quarkus.deployment.annotations.BuildStep;
+import io.quarkus.deployment.builditem.FeatureBuildItem;
+
+class OpenApiGeneratorExtensionProcessor {
+
+    private static final String FEATURE = "openapi-generator-extension";
```
Comment (FroMage) [inline extensions/openapi-generator/deployment/src/main/java/io/quarkus/openapi/generator/deployment/OpenApiGeneratorExtensionProcessor.java:8]: > Features provided by core extensions should be listed in the {@link Feature} enum.

File: extensions/openapi-generator/deployment/src/main/resources/templates/api.qute [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+import org.eclipse.microprofile.rest.client.inject.RegisterRestClient;
+
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.util.List;
+import java.util.Map;
+import javax.ws.rs.*;
+import javax.ws.rs.core.Response;
+import javax.ws.rs.core.MediaType;
+import org.jboss.resteasy.annotations.jaxrs.*;
```
Comment (FroMage) [inline extensions/openapi-generator/deployment/src/main/resources/templates/api.qute:16]: What are those used for?

If they are used, we will need another version of that template for RESTEasy Reactive.

File: extensions/openapi-generator/deployment/src/main/resources/templates/api.qute [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+
+{#insert disableMultipart}
+import org.apache.cxf.jaxrs.ext.multipart.*;
```
Comment (FroMage) [inline extensions/openapi-generator/deployment/src/main/resources/templates/api.qute:19]: This is only for RESTEasy Classic, I think? @michalszynkiewicz we support multipart from the RESTEasy Reactive client, right?

File: extensions/openapi-generator/deployment/src/main/resources/templates/api.qute [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+    {#if op.subresourceOperation}
+    @Path("{op.path}")
+    {/if}
+    {#if op.hasConsumes}
+    @Consumes(\{{#for consume in op.consumes}"{consume.mediaType}"{#if hasNext}, {/if}{/for}\})
+    {/if}
+    {#if op.hasProduces}
+    @Produces(\{{#for produce in op.produces}"{produce.mediaType}"{#if hasNext}, {/if}{/for}\})
+    {/if}
+    public {#if op.returnType}{op.returnType}{#else}void{/if} {op.nickname}({#for p in op.allParams}{#include pathParams.qute param=p/}{#include queryParams.qute param=p/}{#include bodyParams.qute param=p/}{#include formParams.qute param=p/}{#include headerParams.qute param=p/}{#if hasNext}, {/if}{/for});
```
Comment (FroMage) [inline extensions/openapi-generator/deployment/src/main/resources/templates/api.qute:53]: In the case of RESTEasy Reactive (though TBH I don't know if the RESTEasy Classic client supports this as well, I don't recall) we will also need a reactive variant of the method. Either in another type of client `{classname}Reactive` or in the same file, with the following signature:

`public {#if op.returnType}Uni<{op.returnType}>{#else}Uni<Void>{/if}…`

**Review by FroMage — CHANGES_REQUESTED** [review]
Thanks a lot, this is really great that we can use Qute. Will need a review by @mkouba too, for the Qute parts.
I left a few comments.
I wonder how this extension gets called, though? I didn't see how to invoke it, besides programmatically.

**Comment by ricardozanini** [comment]
> I wonder how this extension gets called, though? I didn't see how to invoke it, besides programmatically.

Hey @FroMage! I didn't get there yet. The idea is to generate all the client code in the build phase. The extension will provide the generated code to the platform. All open-api files found on the user's `src/main/resource/openapi` path will have their clients generated.

Of course, the generation process can be fine-tuned by users using proper configuration. For example:

```properties
# multiple classpath paths where the files are located
quarkus.openapi.generator.files=clients,openapi-files,openapi
# prefix for all related configuration for a given openapi interface file
quarkus.openapi.generator.<openapiid>.*
```

The OpenApi ID (`<openapiid>`) is the file name without the extension. So `src/main/resource/openapi/petstore.yaml` properties should be defined as:

```properties
# runtime property
quarkus.openapi.generator.petstore.authtentication.api_key=12345
# see https://openapi-generator.tech/docs/generators/java
# build only property
quarkus.openapi.generator.petstore.apiPackage=org.acme.mypetstore
```
Most configurations from the OpenAPI Generator will be available: 
- https://openapi-generator.tech/docs/globals
- https://openapi-generator.tech/docs/generators/java

wdyt?

One question I have, though. What's the best place to generate such code? In `target/generated-sources`? Or we should add to the user's project directly like in `src/main/gen`? (or it can be customized)

Sorry to take too long, I have other priorities at the moment and I'm sharing my time between other projects as well.

File: extensions/openapi-generator/deployment/src/main/java/io/quarkus/openapi/generator/deployment/template/QuteTemplatingEngineAdapter.java [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+            "formParams.qute",
+            "headerParams.qute",
+            "pathParams.qute",
+            "pojo.qute",
+            "queryParams.qute"
+    };
+    public final Engine engine;
+
+    public QuteTemplatingEngineAdapter() {
+        // TODO: inject from qute buildStep
```
Comment (mkouba) [inline extensions/openapi-generator/deployment/src/main/java/io/quarkus/openapi/generator/deployment/template/QuteTemplatingEngineAdapter.java:33]: When is this class actually used? Qute does not provide a fully initiliazed `Engine` instance during the build phase. But I think it's OK to use a standalone `Engine` instance for this extension only.

**Comment by FroMage** [comment]
> One question I have, though. What's the best place to generate such code? In target/generated-sources? Or we should add to the user's project directly like in src/main/gen? (or it can be customized)

Hah, very good question. I'll ask on the mailing list.

**Comment by FroMage** [comment]
I've asked on the mailing list, but note that GRPC (https://quarkus.io/guides/grpc-getting-started) chose to always generate the stubs as part of the build and picked `target/generated-sources`.

Also note that they put their proto files in `src/main/proto` and not in `src/main/resource/proto`. So perhaps that's the sort of location we should use, with `src/main/openapi`?

Also note that our (server-side) openapi extension places its files in `src/main/resource/META-INF/openapi.yaml` (see https://quarkus.io/guides/openapi-swaggerui#loading-openapi-schema-from-static-files) but this is different because it's for server-side publishing, so I'm not worried if client and server files are in separate locations.

File: extensions/openapi-generator/deployment/src/main/resources/templates/api.qute (hunk 6 above)
Comment (michalszynkiewicz) [reply extensions/openapi-generator/deployment/src/main/resources/templates/api.qute:19]: yes, https://github.com/quarkusio/quarkus/blob/main/docs/src/main/asciidoc/rest-client-reactive.adoc#multipart-form-support

**Comment by michalszynkiewicz** [comment]
How are users meant to use it?

Is it meant to be generated on each build/change?
It sounds to me like a good candidate for `CodeGenProvider`, see https://github.com/quarkusio/quarkus/blob/main/core/deployment/src/main/java/io/quarkus/deployment/CodeGenProvider.java
With this approach you get automatic regeneration of the clients on openapi yaml changes, automatically registered sources root for the generated sources, support for generating test sources and 'normal' sources from `src/test/openapi` and `src/main/openapi` respectively.

File: extensions/openapi-generator/deployment/src/main/java/io/quarkus/openapi/generator/deployment/template/QuteTemplatingEngineAdapter.java (hunk 8 above)
Comment (ricardozanini) [reply extensions/openapi-generator/deployment/src/main/java/io/quarkus/openapi/generator/deployment/template/QuteTemplatingEngineAdapter.java:33]: Oh, this class is instantiated by the OpenAPI Generator tool via service loader. So, are you saying that I can't inject a Qute Engine during the build phase? I guess initializing it manually will be the default option then. Correct?

**Comment by ricardozanini** [comment]
> It sounds to me like a good candidate for CodeGenProvider

Yes, that's the idea. :)

> chose to always generate the stubs as part of the build and picked `target/generated-sources`.

Perfect man!

> Also note that they put their proto files in `src/main/proto` and not in `src/main/resource/proto`. So perhaps that's the sort of location we should use, with `src/main/openapi`?

Not sure since the yaml/json openapi files are resources and not code? I think we should stay with `src/main/resources/META-INF/openapi/client` instead, wdyt?

File: extensions/openapi-generator/deployment/src/main/java/io/quarkus/openapi/generator/deployment/template/QuteTemplatingEngineAdapter.java (hunk 8 above)
Comment (mkouba) [reply extensions/openapi-generator/deployment/src/main/java/io/quarkus/openapi/generator/deployment/template/QuteTemplatingEngineAdapter.java:33]: Yes.

**Comment by FroMage** [comment]
@michalszynkiewicz Huh, what you describe, producing code in `src/main/openapi` is opposite to what GRPC does ATM. Can you check the mailing list discussion and participate there? We need to have an overall consistent strategy for generated code :)

@ricardozanini

> Not sure since the yaml/json openapi files are resources and not code? I think we should stay with `src/main/resources/META-INF/openapi/client` instead, wdyt?

Not unless those files are needed in the application at runtime. Are they?

**Comment by michalszynkiewicz** [comment]
@FroMage I meant producing code **from** src/main/openapi

**Comment by FroMage** [comment]
Oh, sorry, then fine, great :)

**Comment by michalszynkiewicz** [comment]
> @michalszynkiewicz Huh, what you describe, producing code in `src/main/openapi` is opposite to what GRPC does ATM. Can you check the mailing list discussion and participate there? We need to have an overall consistent strategy for generated code :)

So the overall consistent (among grpc and avro :D) strategy for generated code in the quarkus core now is:
- the source for the generation is placed in a subdirectory of `src/main/` (or `src/test` if it's meant only for tests)
- the output is put in a subdirectory of `target/generated-sources` (or `src/generated-test-sources`); and Gradle counterpar of itt for Gradle
- code generation is done in `generate-code` and `generate-code-tests` maven goals, and `QuarkusGenerateCode` Gradle task, the former is added to the `quarkus-maven-plugin` configuration when a project is generated, the latter doesn't need to be added to the config

Code generation is done by implementing `CodeGenProvider` and registering the implementation for the service loader mechanism. CodeGenProviders have access to the deployment classpath.
When running in dev mode, a provider will be retriggered if files it works on change (at the moment one provider works on one directory and one file extension)

File: extensions/openapi-generator/deployment/src/main/resources/templates/api.qute (hunk 5 above)
Comment (ricardozanini) [reply extensions/openapi-generator/deployment/src/main/resources/templates/api.qute:16]: This is just temporary. First thing is bringing all their templates to Qute, then I'll do the proper changes.

**Comment by ricardozanini** [comment]
> Not unless those files are needed in the application at runtime. Are they?

@FroMage, I'm not sure YET. For now, I'll move to `src/main/openapi` then. :)

@michalszynkiewicz that makes sense, many thanks for the inputs.

**Comment by ricardozanini** [comment]
Code generation refactored to `codegen` module inspired by the `grpc` extension. Tomorrow I'll bind the code generation to the build step and hopefully begin integration tests.

Later, I'll tune the templates to be more "Quarkus RESTEasy Like".

File: extensions/openapi-generator/codegen/src/main/java/io/quarkus/openapi/generator/deployment/OpenApiGeneratorCodeGen.java [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+        // TODO: skip codegen?
+        final Path outDir = context.outDir();
+        final Path openApiDir = context.inputDir();
+        try {
+            if (Files.isDirectory(openApiDir)) {
+                try (Stream<Path> openApiFilesPaths = Files.walk(openApiDir)) {
+                    final List<String> openApiFiles = openApiFilesPaths
+                            .filter(Files::isRegularFile)
+                            .map(Path::toString)
+                            .filter(s -> s.endsWith(YAML) || s.endsWith(YML) || s.endsWith(JSON))
```
Comment (michalszynkiewicz) [inline extensions/openapi-generator/codegen/src/main/java/io/quarkus/openapi/generator/deployment/OpenApiGeneratorCodeGen.java:54]: With `inputExtension` returning only `json`, `trigger()` won't be called on `*.yaml` and `*.yml` file changes. 
I think it would be good to change `inputExtension` to a collection or array. I think it should be simple, would require propagating it and changing https://github.com/quarkusio/quarkus/blob/main/core/deployment/src/main/java/io/quarkus/deployment/util/FSWatchUtil.java#L51 to `contains` instead of `equals`

Comment (michalszynkiewicz) [reply extensions/openapi-generator/codegen/src/main/java/io/quarkus/openapi/generator/deployment/OpenApiGeneratorCodeGen.java:54]: an alternative is to do what we have for Avro, separate `CodeGenProvider` for each file extension

Comment (ricardozanini) [reply extensions/openapi-generator/codegen/src/main/java/io/quarkus/openapi/generator/deployment/OpenApiGeneratorCodeGen.java:54]: Yeah, I had this feeling... well, I'll do the Acro way for now ;)

**Comment by ricardozanini** [comment]
@FroMage @michalszynkiewicz the very first raw, basic use case is here. Many thanks for the help!

I was able to generate the code for the PetStore OpenApi spec and integrate it with a Quarkus Application. There is a `QuarkusUnitTest` in the `deployment` module verifying it.

Now I believe I need some help with the extension regarding the generated code. I believe that we would need to register for reflection all the generated classes for native compilation. Can you point to me how do I access the generated code in the `Processor`? 

In order to generate code, I had to add the quarkus plugin to my `deployment` module:

```xml

      <plugin>
        <groupId>io.quarkus</groupId>
        <artifactId>quarkus-maven-plugin</artifactId>
        <extensions>true</extensions>
        <executions>
          <execution>
            <goals>
              <goal>build</goal>
              <goal>generate-code</goal>
              <goal>generate-code-tests</goal>
            </goals>
          </execution>
        </executions>
      </plugin>
```

Is this expected? If not, how can I invoke the `CodeGen` generators from my Test context without it? I took a look at the avro and grpc generators, I could see that you have proto files in the `src/test/proto` directory in the `deployment` module, but I couldn't figure how you're generating the code in the unit tests.

Another point is that the generator has templates for `jackson` only. We can add the option to support jsonb as well, but I'd say to start small now and enhance the extension as we go.

After solving this "issue" with the generator, I'll expose the OpenApi Tool configuration and will work on provide Basic Auth and oauth2 for the generated apis.

To summarize, remaining tasks are:

- [ ] Handle generated code in `deployment` model. Do we need to register for reflection the generated classes? What else do we need?
- [ ] Expose OpenApi Generator Tool configuration via quarkus properties
- [ ] Basic Auth and oAuth2 support
- [ ] Write the Guide

For the future:
- [ ] Handle Multipart 
- [ ] Handle Async Invocations
- [ ] Support jsonb

wdyt?

**Comment by michalszynkiewicz** [comment]
> @FroMage @michalszynkiewicz the very first raw, basic use case is here. Many thanks for the help!

Cool!

> 
> I was able to generate the code for the PetStore OpenApi spec and integrate it with a Quarkus Application. There is a `QuarkusUnitTest` in the `deployment` module verifying it.
> 
> Now I believe I need some help with the extension regarding the generated code. I believe that we would need to register for reflection all the generated classes for native compilation. Can you point to me how do I access the generated code in the `Processor`?

I think the classes you generate should be picked up automatically and `RestClientProcessor` should do any needed registrations itself. But that would have to be tested :)

> 
> In order to generate code, I had to add the quarkus plugin to my `deployment` module:
> 
> ```
>       <plugin>
>         <groupId>io.quarkus</groupId>
>         <artifactId>quarkus-maven-plugin</artifactId>
>         <extensions>true</extensions>
>         <executions>
>           <execution>
>             <goals>
>               <goal>build</goal>
>               <goal>generate-code</goal>
>               <goal>generate-code-tests</goal>
>             </goals>
>           </execution>
>         </executions>
>       </plugin>
> ```
> 
> Is this expected? If not, how can I invoke the `CodeGen` generators from my Test context without it? I took a look at the avro and grpc generators, I could see that you have proto files in the `src/test/proto` directory in the `deployment` module, but I couldn't figure how you're generating the code in the unit tests.
> 

This should not be needed for the deployment module. You may need to add there though it if you want to have tests using your generator in the deployment module.
But, for the specific case of generator, I think having tests in `integration-tests` would make things easier, you wouldn't need to add the quarkus-maven-plugin here. And you can also test native in integration tests.

**Comment by ricardozanini** [comment]
> But, for the specific case of generator, I think having tests in integration-tests would make things easier, you wouldn't need to add the quarkus-maven-plugin here. And you can also test native in integration tests.

That's what I was wondering to do. Many thanks for confirming it! I thought that was some kind of access to the generators in the test context without using the plugin.

I'll work on it and create the configuration scheme for the plugin.

@FroMage are you ok with the plan so far?

**Comment by michalszynkiewicz** [comment]
One more thing, these two probably need to be updated for the gradle code generation to work: 
https://github.com/quarkusio/quarkus/blob/main/devtools/gradle/src/main/java/io/quarkus/gradle/tasks/QuarkusGenerateCode.java#L40-L42

@glefloch knows the details

**Comment by glefloch** [comment]
Right, looking at the pull request, I think it should be : 

````
 public static final String[] CODE_GENERATION_PROVIDER = new String[] { "grpc", "avdl", "avpr", "avsc", "open-api-json" };
 public static final String[] CODE_GENERATION_INPUT = new String[] { "proto", "avro", "openapi" };
````

**Comment by ricardozanini** [comment]
> Right, looking at the pull request, I think it should be :
> 
> ```
>  public static final String[] CODE_GENERATION_PROVIDER = new String[] { "grpc", "avdl", "avpr", "avsc", "open-api-json" };
>  public static final String[] CODE_GENERATION_INPUT = new String[] { "proto", "avro", "openapi" };
> ```

@glefloch we have `open-api-yaml` and `open-api-yml` as well. Should we add them to `CODE_GENERATION_PROVIDER`?

**Comment by glefloch** [comment]
Right @ricardozanini

**Comment by glefloch** [comment]
This will be used by gradle to detect if the task is up to date

**Comment by ricardozanini** [comment]
Rebased, new `integration-tests` module, Gradle support. I should finish the configuration module by the end of this week, whereas this should be ready for the first round of reviews.

**Comment by ricardozanini** [comment]
@FroMage @michalszynkiewicz, how do you guys have access to the `Config` beans in the `CodeGenProvider`? I created a `ConfigGroup` with `CodegenConfig` and `RuntimeConfig` to separate the properties for the respective phases. The first will control how the OpenApi files will be generated, exposing some OpenApi Generator configurations. The latter, users will use to add Authentication, endpoint configuration, and other runtime-related configs.

The problem is that I cannot access the `CodegenConfig` `ConfigRoot` in `CodeGenProvider`. Am I using the wrong phase? I'm assuming the codegen happens during `BUILD_TIME`.

**Comment by ricardozanini** [comment]
As we talked about on Zulip, the codegen configuration will be migrated to the user's POM properties. Relates to https://github.com/quarkusio/quarkus/issues/19389

File: extensions/openapi-generator/deployment/src/main/resources/templates/pojo.qute [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+{#else}
+@JsonIgnoreProperties(ignoreUnknown = true)
+{/if}
+{#if m.description}
+/**
+  * {m.description}
+ **/
+{/if}
+{#include additionalModelTypeAnnotations.qute m=m/}
+public class {m.classname} {#if m.parent}extends {m.parent}{/if}{#if m.serializableModel} implements Serializable{/if} {
```
Comment (FroMage) [inline extensions/openapi-generator/deployment/src/main/resources/templates/pojo.qute:32]: This looks like we're generating getter/setters for the POJOs, but these days we tend to write POJOs using public fields, perhaps we need to change that?

File: extensions/openapi-generator/deployment/src/main/java/io/quarkus/openapi/generator/deployment/wrapper/OpenApiClientGeneratorWrapper.java [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+public class OpenApiClientGeneratorWrapper {
+
+    private static final String VERBOSE = "verbose";
+    private static final String ONCE_LOGGER = "org.openapitools.codegen.utils.oncelogger.enabled";
+
+    private final QuarkusCodegenConfigurator configurator;
+    private final DefaultGenerator generator;
+
+    public OpenApiClientGeneratorWrapper(final String specFilePath, final String outputDir) {
+        // TODO: once we have support, expose these properties as quarkus.open-api-generator.codegen.*
```
[the remark this answers is not in the document]
Comment (ricardozanini) [reply extensions/openapi-generator/deployment/src/main/java/io/quarkus/openapi/generator/deployment/wrapper/OpenApiClientGeneratorWrapper.java:29]: I'm working on it. The properties from the build stage are not being read. I'll use maven properties instead. :(

File: extensions/openapi-generator/deployment/src/main/resources/templates/pojo.qute (hunk 10 above)
Comment (ricardozanini) [reply extensions/openapi-generator/deployment/src/main/resources/templates/pojo.qute:32]: Yeah, these are accessors. I can get rid of them and change to public fields. np.

File: extensions/openapi-generator/deployment/src/main/java/io/quarkus/openapi/generator/deployment/codegen/OpenApiGeneratorCodeGenBase.java [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+            }
+        } catch (IOException e) {
+            throw new CodeGenException("Failed to generate java files from OpenApi files in " + openApiDir.toAbsolutePath(), e);
+        }
+        return false;
+    }
+
+    private String escapeWhitespace(String path) {
+        if (OS.determineOS() == OS.LINUX) {
+            return path.replace(" ", "\\ ");
```
[the remark this answers is not in the document]
Comment (ricardozanini) [reply extensions/openapi-generator/deployment/src/main/java/io/quarkus/openapi/generator/deployment/codegen/OpenApiGeneratorCodeGenBase.java:65]: I copied this code from the `grpc` extension. Although I suspect the `Path` API could handle this use case for us. :thinking:

File: extensions/openapi-generator/deployment/src/main/java/io/quarkus/openapi/generator/deployment/wrapper/OpenApiClientGeneratorWrapper.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (michalszynkiewicz) [reply extensions/openapi-generator/deployment/src/main/java/io/quarkus/openapi/generator/deployment/wrapper/OpenApiClientGeneratorWrapper.java:29]: `application.properties` are available in code generator now

[the remark this answers is not in the document]
Comment (ricardozanini) [reply extensions/openapi-generator/deployment/src/main/java/io/quarkus/openapi/generator/deployment/wrapper/OpenApiClientGeneratorWrapper.java:29]: @michalszynkiewicz I'm trying to inject my configuration to my `CodeGenProvider` implementation but it looks like the engine is not injecting it. I'm receiving NPEs. How do I access the build time configuration properties from a `CodeGenProvider` impl? Can you help?

[the remark this answers is not in the document]
Comment (michalszynkiewicz) [reply extensions/openapi-generator/deployment/src/main/java/io/quarkus/openapi/generator/deployment/wrapper/OpenApiClientGeneratorWrapper.java:29]: You cannot inject it, you can get the MP Config `org.eclipse.microprofile.config.Config` by calling `context.config()`
